changes/2020-07-15_specify-behavior-for-raw-aes-keyring-for-invalid-EC/change.md

```diff
@@ -0,0 +1,99 @@
+| ---------- | ------------------------------------------------------------------------------------- |
+| C          | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-c)                   |
+| Javascript | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-javascript) |
+
+## Definitions
+
+- "serialize", "serialization", and "serializability" all refer to
+  [message header serialization of AAD key value pairs](../../data-format/message-header.md#key-value-pairs)
+
+### Conventions used in this document
+
+The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
+in this document are to be interpreted as described in
+[RFC 2119](https://tools.ietf.org/html/rfc2119).
+
+## Summary
+
+The Raw AES Keyring must serialize the encryption context for use as AAD.
+The specification did not describe how the keyring should behave when given an unserializable encryption context.
+
+With this change, the Raw AES Keyring MUST attempt to serialize the encryption context,
+prior to attempting encryption/decryption.
+If the Raw AES Keyring cannot serialize the encryption context, it MUST fail.
+
+## Out of Scope
+
+- All keyrings aside from the Raw AES Keyring
+- The encryption context serialization format
+
+## Motivation
+
+The C and Javascript implementations of the Raw AES Keyring check for serializability at different points in OnDecrypt.
+This has the potential to lead to different behaviors under specific circumstances.
+
+## Drawbacks
+
+Since we attempt to serialize the encryption context before attempting any decryption,
+the Raw AES Keyring may attempt to serialize an encryption context in cases where no EDK matches the keyring.
+This scenario results in slower operation when compared to the current Javascript implementation,
+which only checks serializability when attempting to decrypt an EDK.
+
+## Security Implications
+
+This change SHOULD NOT have any security implications.
+
+## Operational Implications
+
+This change will improve behavioral consistency between Raw AES Keyring implications.
+This should improve customer experience and reduce support engagements.
+
+## Guide-level Explanation
+
+Serialize the encryption/decryption materials' encryption context before attempting encryption/decryption.
+If serialization fails, the OnEncrypt/OnDecrypt operation MUST fail.
+When encrypting/decryption the datakey, use the already serialized encryption context as the AAD.
+
+## Reference-level Explanation
+
+### OnEncrypt
+
+Before encrypting the plaintext datakey, OnEncrypt MUST attempt to serialize the encryption context.
+If serialization fails, OnEncrypt MUST fail.
+This is not a new change, but is more explicit now.
+
+### OnDecrypt
+
+Before iterating through the list of EDKs, OnDecrypt MUST attempt to serialize the encryption context.
```

Comment:
Nit: If there are no EDKs in the list that match this name/namspace pair then I don't think the keyring should be required to successfully serialize.
I think that this also means that the spec should be updated.
Because I don't think that
"The keyring MUST perform the following actions on each encrypted data key in the input encrypted data key list, serially, until it successfully decrypts one."
Is correct.
For example it should NOT be attempting to decrypt AWS KMS ciphertext blobs.

Reply:
> Nit: If there are no EDKs in the list that match this name/namspace pair then I don't think the keyring should be required to successfully serialize.

I see the logic behind that (if you look back in the commits on this branch, you'll see that's what I originally wrote). The way I see it though is an unserializable EC means we're in bad enough state that there's no reason to let decryption go on.

> For example it should NOT be attempting to decrypt AWS KMS ciphertext blobs

I believe the spec already accounts for this:

> The keyring MUST attempt to decrypt the encrypted data key if and only if the following is true:
> - ...
> - the key name obtained from the encrypted data key's key provider information has a value equal to this keyring's [key name](#key-name).
> - the key provider ID of the encrypted data key has a value equal to this keyring's [key namespace](#key-namespace).

With the change in #140, the Raw AES Keyring should never attempt to decrypt KMS EDKs.

Reply:
The CMM can modify the encryption context.
So if the CMM could modifies the EC in such a way that the Raw keyring would fail.
On the other hand, since we now deny noop keyrings
the raw AES Keyring MUST fail if there are no EDKS
and this strange interaction is better communicated your way.